src/lib/Simplify.hs

```diff
@@ -75,53 +75,50 @@ simplifyAtom atom = case atom of
   -- We don't simplify body of lam because we'll beta-reduce it soon.
-  Lam _ -> substEmbed atom
-  Pi  _ -> substEmbed atom
+  Lam _ -> substEmbedR atom
+  Pi  _ -> substEmbedR atom
   Con (AnyValue (TabTy v b)) -> TabValA v <$> mkAny b
   Con (AnyValue (PairTy a b))-> PairVal <$> mkAny a <*> mkAny b
   Con (AnyValue (SumTy l r)) -> do
     Con <$> (SumCon <$> mkAny (TC $ BaseType $ Scalar BoolType) <*> mkAny l <*> mkAny r)
   Con con -> Con <$> mapM simplifyAtom con
-  TC tc -> TC <$> mapM substEmbed tc
-  Eff eff -> Eff <$> substEmbed eff
-  where mkAny t = Con . AnyValue <$> substEmbed t >>= simplifyAtom
-
--- Unlike `substEmbed`, this simplifies under the binder too.
-simplifyLam :: Atom -> SimplifyM (Atom, Maybe (Atom -> SimplifyM Atom))
-simplifyLam atom = substEmbed atom >>= (dropSub . simplifyLam')
-
-simplifyLam' :: Atom -> SimplifyM (Atom, Maybe (Atom -> SimplifyM Atom))
-simplifyLam' (Lam (Abs b (arr, body))) = do
-  b' <- mapM substEmbed b
-  if isData (getType body)
-    then liftM (,Nothing) $ do
-      buildDepEffLam b' (\x -> extendR (b@>x) $ substEmbed arr)
-                        (\x -> extendR (b@>x) $ simplifyBlock body)
-    else do
-      (lam, recon) <- buildLamAux b'
-        ( \x -> extendR (b@>x) $ substEmbed arr)
-        $ \x -> extendR (b@>x) $ do
-          (body', (scope, decls)) <- embedScoped $ simplifyBlock body
-          mapM_ emitDecl decls
-          return $ separateDataComponent scope body'
-      return $ (lam, Just recon)
-simplifyLam' atom = error $ "Not a lambda: " ++ pprint atom
-
-simplifyBinaryLam :: Atom -> SimplifyM Atom
-simplifyBinaryLam atom = substEmbed atom >>= (dropSub . simplifyBinaryLam')
-
-simplifyBinaryLam' :: Atom -> SimplifyM Atom
-simplifyBinaryLam' (BinaryFunVal b1 b2 eff body) = do
-  b1' <- mapM substEmbed b1
-  buildLam b1' PureArrow $ \x1 ->
-    extendR (b1'@>x1) $ do
-      b2' <- mapM substEmbed b2
-      buildDepEffLam b2'
-        (\x2 -> extendR (b2'@>x2) $ substEmbed (PlainArrow eff))
-        (\x2 -> extendR (b2'@>x2) $ simplifyBlock body)
-simplifyBinaryLam' atom = error $ "Not a binary lambda: " ++ pprint atom
+  TC tc -> TC <$> mapM substEmbedR tc
+  Eff eff -> Eff <$> substEmbedR eff
+  where mkAny t = Con . AnyValue <$> substEmbedR t >>= simplifyAtom
+
+
+-- `Nothing` is equivalent to `Just return` but we can pattern-match on it
+type Recon m a = Maybe (a -> m a)
+
+simplifyLam :: Atom -> SimplifyM (Atom, Recon SimplifyM Atom)
+simplifyLam = simplifyLams 1
+
+simplifyBinaryLam :: Atom -> SimplifyM (Atom, Recon SimplifyM Atom)
+simplifyBinaryLam = simplifyLams 2
+
+-- Unlike `substEmbedR`, this simplifies under the binder too.
+simplifyLams :: Int -> Atom -> SimplifyM (Atom, Recon SimplifyM Atom)
+simplifyLams numArgs lam = do
+  lam' <- substEmbedR lam
+  dropSub $ simplifyLams' numArgs mempty $ Block [] $ Atom lam'
+
+simplifyLams' :: Int -> Scope -> Block -> SimplifyM (Atom, Recon SimplifyM Atom)
+-- this case is an optimization
```

Comment:
didn't you mean to put this comment at the case below (the one for an empty block)?

Reply:
I think it's just stale. If anything, it was probably for the `isData` case. I'll delete it.